### PR TITLE
[OptimizeInstructions] Optimize memory.fill with constant arguments

### PR DESCRIPTION
**Pull request by MaxGraey**
This is reland of #3071

Do similar optimizations as for #3038 but for memory.fill.
`memory.fill(d, v, 0)` ==> `{ drop(d), drop(v) }` only with `ignoreImplicitTraps` or `trapsNeverHappen`
`memory.fill(d, v, 1)` ==> `store8(d, v)`

Further simplifications could be done only if v (value) is constant. Otherwise it increase binary size:
`memory.fill(d, C, 1)` ==> `store8(d, (C & 0xFF))`
`memory.fill(d, C, 2)` ==> `store16(d, (C & 0xFF) * 0x0101)`
`memory.fill(d, C, 4)` ==> `store32(d, (C & 0xFF) * 0x01010101)`
`memory.fill(d, C, 8)` ==> `store64(d, (C & 0xFF) * 0x0101010101010101)`
`memory.fill(d, C, 16)` ==> `store128(d, i8x16.splat(C & 0xFF))`

### CODE REVIEW

**Comment by MaxGraey** [comment]
@kripken it seems `--ignore-implicit-traps` doesn't work in Lit tests anymore?

**Comment by kripken** [comment]
We do have tests for `--ignore-implicit-traps` (`grep ignore-imp test/lit/passes/*`) which have not regressed so I am not aware of a general problem. What specifically do you see?

**Comment by MaxGraey** [comment]
In this tests all expressions should lower to drops:
https://github.com/WebAssembly/binaryen/pull/4130/files#diff-ff9e85ef78b1e2410aa24601d96db7422161b74cc5522a44435ff51d08e5e40fR764

But in this test it happens even though it should:
https://github.com/WebAssembly/binaryen/pull/3071/files#diff-a564949105908d8bfc9ee0442513a3b0435126bb50c5c32aabeb4b18e4fd6cc5

**Comment by MaxGraey** [comment]
Alright I figured out what's problem. `optimize-instructions-ignore-traps.wast` had a missing `--optimize-instruction` flag

**Comment by MaxGraey** [comment]
@tlively could you take a look at this ancient PR, please?

File: src/passes/OptimizeInstructions.cpp [hunk 1]
```diff
@@ -2851,6 +2861,103 @@ struct OptimizeInstructions
...
+            if (options.shrinkLevel == 0) {
+              if (getModule()->features.hasSIMD()) {
+                return builder.makeStore(
+                  16, // bytes
+                  0,  // offset
+                  1,  // align
+                  memFill->dest,
+                  builder.makeUnary(
+                    SplatVecI8x16,
+                    builder.makeConst<uint8_t>((uint8_t)byteValue)),
```
Comment (tlively) [inline src/passes/OptimizeInstructions.cpp:2938]: It would be better to use a `v128.const` here. Also, if SIMD is not enabled, it might be worth splitting this into two i64 stores.

File: test/lit/passes/optimize-instructions-ignore-traps.wast [hunk 2]
```diff
@@ -1,5 +1,5 @@
 ;; NOTE: Assertions have been generated by update_lit_checks.py and should not be edited.
-;; RUN: wasm-opt %s --all-features --optimize-level=2 --ignore-implicit-traps -S -o - \
+;; RUN: wasm-opt %s --all-features --optimize-instructions --optimize-level=2 --ignore-implicit-traps -S -o - \
```
Comment (tlively) [inline test/lit/passes/optimize-instructions-ignore-traps.wast:2]: Can you add these tests in a new file to avoid making unrelated changes to other tests in this file?

Comment (MaxGraey) [reply test/lit/passes/optimize-instructions-ignore-traps.wast:2]: Basically this lit required this `--optimize-instructions`. It was miss during porting I guess

Comment (tlively) [reply test/lit/passes/optimize-instructions-ignore-traps.wast:2]: Oh, I see. It's concerning that we accidentally dropped the key option for the test. I'll take a look at what went wrong there.

File: test/lit/passes/optimize-instructions-ignore-traps.wast [hunk 3]
```diff
@@ -759,4 +754,65 @@
...
+  (memory.fill ;; drops
+    (i32.load (i32.const -1))
+    (i32.const 0)
+    (i32.const 0)
+  )
+
+  (memory.fill ;; skip
+    (i32.const 0)
+    (i32.const 0)
+    (i32.load (i32.const -1))
```
Comment (tlively) [inline test/lit/passes/optimize-instructions-ignore-traps.wast:815]: Could this be a `local.get $sz`? I notice we haven't used the $sz parameter, and this load from a negative index looks like it should be testing something about the load, but actually the point is just that it's not a constant, right?

**Review by tlively — COMMENTED** [review]
LGTM with that last potential fix.

File: test/lit/passes/optimize-instructions-ignore-traps.wast (hunk 3 above)
Comment (MaxGraey) [reply test/lit/passes/optimize-instructions-ignore-traps.wast:815]: Yeah, makes sense

**Review by tlively — APPROVED** [review]
Thanks!

**Comment by kripken** [comment]
Looks like the fuzzer found a problem here:
```wat
(module
 (type $i32_i32_i32_=>_none (func (param i32 i32 i32)))
 (memory $0 (shared 16 17))
 (func $0 (param $0 i32) (param $1 i32) (param $2 i32)
  (memory.fill
   (i32.const 0)
   (i32.const 0)
   (i32.const 16)
  )
 )
)
```
```
BINARYEN_PASS_DEBUG=1 wasm-opt a.wat -O1 --enable-threads --enable-bulk-memory
```
```
[wasm-validator error in function 0] expression seen more than once in the tree in 0 on 0x563be249dda8
, on 
(i32.const 0)
```
Looks like you are reusing a const node in 2 places.

Separate issue: Very oddly, this does not happen if I replace the `--enable-*` with `-all`. @tlively should those not be synonyms?

File: src/passes/OptimizeInstructions.cpp [hunk 4]
```diff
@@ -2919,6 +2929,121 @@ struct OptimizeInstructions
...
+                  offset,
+                  align,
+                  memFill->dest,
+                  builder.makeConst<uint64_t>(value * 0x0101010101010101ULL),
+                  Type::i64),
+                builder.makeStore(
+                  8,
+                  offset + 8,
+                  align,
+                  memFill->dest,
```
Comment (kripken) [inline src/passes/OptimizeInstructions.cpp:3026]: I suspect reusing `memFill->dest` is the bug here.

Comment (tlively) [reply src/passes/OptimizeInstructions.cpp:3026]: Oops, sorry for missing that in review. @kripken do you think it's worth reverting this while that is fixed?

Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:3026]: The fix should be quick, I think, if we agree this is the problem? Based on time zone @MaxGraey may not be available atm but I can make a PR unless you've already started.

Comment (MaxGraey) [reply src/passes/OptimizeInstructions.cpp:3026]: Sorry @kripken. It was added in last moment without proper testings

Comment (kripken) [reply src/passes/OptimizeInstructions.cpp:3026]: No worries @MaxGraey . But in the future let's not add things in the last moment, and also every thing added should have test coverage.

**Comment by kripken** [comment]
@tlively did you have time to look at

> Separate issue: Very oddly, this does not happen if I replace the `--enable-*` with `-all` [..] should those not be synonyms?

?

If that has a bug then I worry we have all sorts of incomplete test coverage...

**Comment by tlively** [comment]
Replacing `--enable-threads --enable-bulk-memory` with `-all` additionally enables SIMD, which caused this optimization to use a different code path and avoid the error, IIRC. Was there another issue that this doesn't explain?

**Comment by kripken** [comment]
@tlively Ah, that makes sense now! Thanks for looking at this again.
